### ISSUE TITLE
TabNet training not working with PyTorch version of Ludwig

### ISSUE THREAD

**Issue by dantreiman** [ludwig-ai/ludwig#1605]
Tested on MacBook Pro i9 and M1.

TabNet does not appear to be training in Ludwig using PyTorch.  Tried higgs boson example as well as mushroom edibility.

Observed:

- Loss does not decrease between epochs
- AUC is stuck at 0.5 or close

Desired Behavior:

- Training TabNet with Torch should converge; loss should decrease as model trains
- Results with PyTorch should match Tensorflow reference results

For example, on the mushroom edibility experiment in the [experiments repo](https://github.com/ludwig-ai/experiments)
`experiments/automl/validation/mushroom_edibility/train_tabnet_sanity_laptop.py`

Loss after 1 epoch (Ludwig 0.4, TensorFlow)
```
Took 42.9332s
╒═════════╤════════╤════════════╤═════════════╕
│ class   │   loss │   accuracy │   hits_at_k │
╞═════════╪════════╪════════════╪═════════════╡
│ train   │ 0.6961 │     0.6016 │      1.0000 │
╘═════════╧════════╧════════════╧═════════════╛
```

Loss after 1 epoch (Ludwig 0.5.dev0, Torch)
```
Took 1.5725s
╒═════════╤════════╤════════════╕
│ class   │   loss │   accuracy │
╞═════════╪════════╪════════════╡
│ train   │ 0.9712 │     0.5032 │
╘═════════╧════════╧════════════╛
```

Run logs for comparison:
[Ludwig_0.4_tensorflow_mushroom.log](https://github.com/ludwig-ai/ludwig/files/7731209/Ludwig_0.4_tensorflow_mushroom.log)

[Ludwig_0.5.dev0_pytorch_mushroom.log](https://github.com/ludwig-ai/ludwig/files/7731210/Ludwig_0.5.dev0_pytorch_mushroom.log)

**Comment by dantreiman** [comment]
See also: [Training differences between PyTorch and Tensorflow versions of Ludwig](https://github.com/ludwig-ai/ludwig/issues/1613)

**Comment by tgaddair** [comment]
Config:

```
output_features:
  - name: class
    type: category
input_features:
  - name: cap-shape
    type: category
  - name: cap-surface
    type: category
  - name: cap-color
    type: category
  - name: bruises?
    type: category
  - name: odor
    type: category
  - name: gill-attachment
    type: category
  - name: gill-spacing
    type: category
  - name: gill-size
    type: category
  - name: gill-color
    type: category
  - name: stalk-shape
    type: category
  - name: stalk-root
    type: category
  - name: stalk-surface-above-ring
    type: category
  - name: stalk-surface-below-ring
    type: category
  - name: stalk-color-above-ring
    type: category
  - name: stalk-color-below-ring
    type: category
  - name: veil-type
    type: category
  - name: veil-color
    type: category
  - name: ring-number
    type: category
  - name: ring-type
    type: category
  - name: spore-print-color
    type: category
  - name: population
    type: category
  - name: habitat
    type: category
combiner:
  type: tabnet
  size: 24  # N_a
  output_size: 26  # N_d
  sparsity: 0.000001  # lambda_sparse
  bn_virtual_divider: 8  # factor to divide batch_size B to get B_v from the paper
  bn_momentum: 0.6  # m_B
  num_steps: 5  # N_steps
  relaxation_factor: 1.5  # gamma
  use_keras_batch_norm: false
  bn_virtual_bs: 512  # B_v 
training:
  batch_size: 4096 # B
  eval_batch_size: 4096 # 65536 131072 262144 524288
  early_stop: 20
  learning_rate: 0.02
  optimizer:
    type: adam
  decay: true
  decay_steps: 20000
  decay_rate: 0.9
  staircase: true
  regularization_lambda: 1
  validation_field: class
```

Script:

```
import logging

from ludwig.api import LudwigModel
from ludwig.datasets import mushroom_edibility

model = LudwigModel(
    config='config.yaml',
    logging_level=logging.INFO,
    backend="local",
)

mushroom_edibility_df = mushroom_edibility.load()
model.train(
    dataset=mushroom_edibility_df,
    experiment_name='mushroom_edibility_tabnet',
    model_name='mushroom_edibility_tabnet',
    skip_save_processed_input=True
)
```

**Comment by tgaddair** [comment]
@dantreiman after digging into this, I don't believe this is a generic issue with the PyTorch version, rather I think it is something specific to the TabNet implementation. While PyTorch generally shows differences in training behavior, the problem here is more fundamental, that nothing appears to be learned.

**Comment by tgaddair** [comment]
We can compare against [pytorch_tabnet](https://github.com/dreamquark-ai/tabnet/blob/develop/pytorch_tabnet/tab_network.py#L487), and possibly integrate with this framework as an optional dependency.

**Comment by dantreiman** [comment]
This graph is from training forest cover for 3 epochs with TabNet:

PyTorch is orange, tensorflow is blue.  Note for PyTorch loss is going down but the accuracy is not improving.

![Screen Shot 2022-01-07 at 2 24 50 PM](https://user-images.githubusercontent.com/687280/148615205-387cba53-82fd-4b9d-a389-2dd9691c1e73.png)

**Comment by dantreiman** [comment]
This issue is not limited to TabNet, and seems to appear with some datasets regardless of model choice.

Here is forest cover, trained for 10 epochs, tabnet and concat conbiner, tensorflow and pytorch:

![Screen Shot 2022-01-07 at 2 49 55 PM (2)](https://user-images.githubusercontent.com/687280/148617295-9293a1f0-70b7-451d-87a3-fd3e11ceb0d6.png)

**Comment by dantreiman** [comment]
Progress!  Sparsity loss was not being added to the combined train loss, and L2 regularization was being applied by default (which it shouldn't be).  Fixed in `dantreiman/tabular_data_training_issues`.  After these fixes.... accuracy goes up as loss goes down so its better than before.  However, TabNet in PyTorch is still worse than Tensorflow, and the epoch_loss doesn't match the step_training_loss.

![Screen Shot 2022-01-07 at 4 44 08 PM](https://user-images.githubusercontent.com/687280/148624335-b93f1962-0650-44d5-8713-e5dbf6ab6ee2.png)

**Comment by dantreiman** [comment]
I think this is fixed in PR #1663

With this PR, TabNet training history in PyTorch matches Tensorflow reasonable well:

<img width="1728" alt="Screen Shot 2022-01-09 at 3 23 13 PM" src="https://user-images.githubusercontent.com/687280/148706225-c6fd67b1-0f5d-4309-9083-7d9b85a193aa.png">
